include/sys/cbprintf_internal.h

```diff
@@ -0,0 +1,270 @@
+	*_Generic((arg) + 0, \
+		char : (int *)buf, \
+		unsigned char: (int *)buf, \
+		short : (int *)buf, \
+		unsigned short : (int *)buf, \
+		int : (int *)buf, \
+		unsigned int : (unsigned int *)buf, \
+		long : (long *)buf, \
+		unsigned long : (unsigned long *)buf, \
+		long long : (long long *)buf, \
+		unsigned long long : (unsigned long long *)buf, \
+		float : (double *)buf, \
+		double : (double *)buf, \
+		long double : (long double *)buf, \
+		default : \
+			(const void **)buf) = arg
+
+/** @brief Return alignment needed for given argument.
+ *
+ * @param _arg Argument
+ *
+ * @return Alignment in bytes.
+ */
+#define Z_CBPRINTF_ALIGNMENT(_arg) \
+	MAX(_Generic((_arg) + 0, \
+		float : VA_STACK_ALIGN(double), \
+		double : VA_STACK_ALIGN(double), \
+		long double : VA_STACK_ALIGN(long double), \
+		long long : VA_STACK_ALIGN(long long), \
+		unsigned long long : VA_STACK_ALIGN(long long), \
+		default : \
+			__alignof__((_arg) + 0)), VA_STACK_MIN_ALIGN)
+
+/** @brief Detect long double variable.
+ *
+ * @param x Argument.
+ *
+ * @return 1 if @p x is a long double, 0 otherwise.
+ */
+#define Z_CBPRINTF_IS_LONGDOUBLE(x) \
+	_Generic((x) + 0, long double : 1, default : 0)
+
+/** @brief Safely package arguments to a buffer.
+ *
+ * Argument is put into the buffer if capable buffer is provided. Length is
+ * incremented even if data is not packaged.
+ *
+ * @param _buf buffer.
+ *
+ * @param _idx index. Index is postincremented.
+ *
+ * @param _max maximum index (buffer capacity).
+ *
+ * @param _arg argument.
+ */
+#define Z_CBPRINTF_PACK_ARG2(_buf, _idx, _max, _arg) do { \
+	BUILD_ASSERT(!((sizeof(double) < VA_STACK_ALIGN(long double)) && \
+			Z_CBPRINTF_IS_LONGDOUBLE(_arg) && \
+			!IS_ENABLED(CONFIG_CBPRINTF_STATIC_PACKAGE_LONGDOUBLE)),\
+			"Packaging of long double not enabled in Kconfig."); \
+	while (_idx % Z_CBPRINTF_ALIGNMENT(_arg)) { \
+		_idx += sizeof(int); \
+	}; \
+	uint32_t _arg_size = Z_CBPRINTF_ARG_SIZE(_arg); \
+	if (_buf && _idx < _max) { \
+		Z_CBPRINTF_STORE_ARG(&_buf[_idx], _arg); \
+	} \
+	_idx += _arg_size; \
+} while (0)
+
+/** @brief Package single argument.
+ *
+ * Macro is called in a loop for each argument in the string.
+ *
+ * @param arg argument.
+ */
+#define Z_CBPRINTF_PACK_ARG(arg) \
+	Z_CBPRINTF_PACK_ARG2(__pbuf, __plen, __pmax, arg)
+
+/** @brief Package descriptor.
+ *
+ * @param len Package length.
+ *
+ * @param str_cnt Number of strings stored in the package.
+ */
+struct z_cbprintf_desc {
+	uint8_t len;
+	uint8_t str_cnt;
+};
+
+/** @brief Package header. */
+union z_cbprintf_hdr {
+	struct z_cbprintf_desc desc;
+	void *raw;
+};
+
+/** @brief Statically package a formatted string with arguments.
+ *
+ * @param buf buffer. If null then only length is calculated.
+ *
+ * @param _len buffer capacity on input, package size on output.
+ *
+ * @param ... String with variable list of arguments.
+ */
+#define Z_CBPRINTF_STATIC_PACKAGE(buf, _len, ... /* fmt, ... */) \
```

Comment:
Do we have existing uses of reference parameters in macros?  It's not a natural idiom in C.

Reply:
that was initially mimicking function call which was taking length by reference. I will remove it.

Reply:
given https://github.com/zephyrproject-rtos/zephyr/pull/32863#discussion_r591510391 i'm not sure that i understand now. So since `_len` is input and output should it be passed by pointer?

Reply:
I'm saying that people generally won't expect macro parameters to be both input and output, because you can't do that in functions.  The way the behavior is accomplished in functions is to accept a pointer.  To use a different idiom for in+out parameters in macros just because they're implemented as textual substitution creates an opportunity for confusion.

This is my opinion, I'm not blocking on it.

Reply:
updated to have `inlen` and `outlen` parameters.